File: backend/app/services/blog_service.py
from __future__ import annotations

from typing import Any

from bson import ObjectId
from bson.errors import InvalidId

from app.core.exceptions import BlogDomainError
from app.repositories.blog_repo import BlogRepo


class BlogService:
    def __init__(self, repo: BlogRepo) -> None:
        self.repo = repo

    @staticmethod
    def _serialize_datetime(value) -> str | None:
        return value.isoformat() if value else None

    @staticmethod
    def _serialize_post(post) -> dict:
        return {
            "_id": str(post.id),
            "title": post.title,
            "body": post.body,
            "summary": post.summary,
            "cover": post.cover,
            "tags": post.tags or [],
            "is_pinned": post.is_pinned,
            "created_at": BlogService._serialize_datetime(post.created_at),
            "updated_at": BlogService._serialize_datetime(post.updated_at),
        }

    @staticmethod
    def _parse_object_id(value: str, message: str) -> ObjectId:
        try:
            return ObjectId(value)
        except InvalidId as exc:
            raise BlogDomainError(message, 400) from exc

    @staticmethod
    def _clean_tags(tags: list[str] | None) -> list[str]:
        """Dedup + strip + truncate; drop empties."""
        cleaned: list[str] = []
        if not tags:
            return cleaned
        for tag in tags:
            value = tag.strip()
            if value and value not in cleaned:
                cleaned.append(value[:50])
        return cleaned

    async def get_blogs(self, page: int, search: str | None) -> dict:
        per_page = 10
        posts, total = await self.repo.list_posts(
            page=page,
            per_page=per_page,
            search=search,
        )

        pages = (total + per_page - 1) // per_page if per_page > 0 else 0
        tag_counts = await self.repo.aggregate_tag_counts()

        post_dicts: list[dict[str, Any]] = [
            self._serialize_post(post) for post in posts
        ]

        return {
            "posts": post_dicts,
            "tags": tag_counts,
            "pagination": {
                "page": page,
                "per_page": per_page,
                "total": total,
                "pages": pages,
                "has_prev": page > 1,
                "has_next": page < pages,
                "prev_num": page - 1 if page > 1 else None,
                "next_num": page + 1 if page < pages else None,
            },
        }

    async def get_blog_post(self, post_id: str | None) -> dict:
        if not post_id:
            raise BlogDomainError(
                "Either _id or post_id parameter is required",
                400,
            )

        self._parse_object_id(post_id, "Invalid blog post ID")
        post = await self.repo.get_post_by_id(post_id)
        if not post:
            raise BlogDomainError("Blog post not found", 404)

        return self._serialize_post(post)

    async def list_tags(self) -> list[dict[str, int]]:
        return await self.repo.aggregate_tag_counts()

    async def get_posts_by_tag(
        self,
        tag: str,
        *,
        page: int = 1,
        per_page: int = 10,
    ) -> dict:
        tag = tag.strip()
        if not tag:
            raise BlogDomainError("Tag is required", 400)

        posts, total = await self.repo.list_posts_by_tag(
            tag,
            page=page,
            per_page=per_page,
        )
        post_dicts = [self._serialize_post(post) for post in posts]
        return {"posts": post_dicts, "tag": tag, "total": total}

File: go-backend/internal/service/blog_service.go
package service

import (
	"context"
	"errors"
	"strings"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"

	blogerrs "github.com/KanoCifer/kuroome-blog/internal/domain/blog/errs"
	"github.com/KanoCifer/kuroome-blog/internal/dto"
	"github.com/KanoCifer/kuroome-blog/internal/mongo/document"
)

// BlogRepositoryer 定义博客读表面对 posts 集合的读写契约。
type BlogRepositoryer interface {
	ListPosts(ctx context.Context, page, perPage int, search string) ([]document.Post, int64, error)
	AggregateTagCounts(ctx context.Context) ([]document.TagCount, error)
	GetPostByID(ctx context.Context, id string) (*document.Post, error)
	IncrementViews(ctx context.Context, id string) error
	IncrementLikes(ctx context.Context, id string) (int, error)
	ListPostsByTag(ctx context.Context, tag string, page, perPage int) ([]document.Post, int64, error)
}

// Bloger 定义博客读表面的用例契约。

type BlogService struct {
	repo BlogRepositoryer
}

func NewBlogService(repo BlogRepositoryer) *BlogService {
	return &BlogService{repo: repo}
}

func pagination(page, perPage, total int) dto.Pagination {
	pages := 0
	if perPage > 0 {
		pages = (total + perPage - 1) / perPage
	}
	prev, next := (*int)(nil), (*int)(nil)
	if page > 1 {
		v := page - 1
		prev = &v
	}
	if page < pages {
		v := page + 1
		next = &v
	}
	return dto.Pagination{
		Page:    page,
		PerPage: perPage,
		Total:   total,
		Pages:   pages,
		HasPrev: page > 1,
		HasNext: page < pages,
		PrevNum: prev,
		NextNum: next,
	}
}

// ListPosts 分页列出博客（含标签聚合）—— 与 Python get_blogs 对齐。
func (s *BlogService) ListPosts(ctx context.Context, page int, search string) (*dto.BlogListResponse, error) {
	if page < 1 {
		page = 1
	}
	const perPage = 10

	posts, total, err := s.repo.ListPosts(ctx, page, perPage, search)
	if err != nil {
		return nil, err
	}

	tagCounts, err := s.repo.AggregateTagCounts(ctx)
	if err != nil {
		return nil, err
	}

	tags := make([]dto.TagResponse, 0, len(tagCounts))
	for _, tc := range tagCounts {
		tags = append(tags, dto.TagResponse{Name: tc.Name, Count: tc.Count})
	}

	return &dto.BlogListResponse{
		Posts:      dto.ToPostList(posts),
		Tags:       tags,
		Pagination: pagination(page, perPage, int(total)),
	}, nil
}

// GetPost 按 ID 获取单篇博客 —— 与 Python get_blog_post 对齐。
func (s *BlogService) GetPost(ctx context.Context, id string) (*dto.PostResponse, error) {
	if id == "" {
		return nil, blogerrs.ErrInvalidPostID
	}
	if _, err := bson.ObjectIDFromHex(id); err != nil {
		return nil, blogerrs.ErrInvalidPostID
	}

	post, err := s.repo.GetPostByID(ctx, id)
	if err != nil {
		if errors.Is(err, mongo.ErrNoDocuments) {
			return nil, blogerrs.ErrPostNotFound
		}
		return nil, err
	}
	out := dto.ToPostResponse(*post)
	return &out, nil
}

// IncrementViews 原子递增单篇文章的浏览量。
// 调用方以 fire-and-forget goroutine 触发，不阻塞读取路径。
func (s *BlogService) IncrementViews(ctx context.Context, id string) error {
	if id == "" {
		return blogerrs.ErrInvalidPostID
	}
	return s.repo.IncrementViews(ctx, id)
}

// LikePost 原子递增单篇文章的喜欢数并返回递增后的值。
// 一次性表态：调用方（handler / 客户端）负责幂等，服务端不做重复判定。
func (s *BlogService) LikePost(ctx context.Context, id string) (int, error) {
	if id == "" {
		return 0, blogerrs.ErrInvalidPostID
	}
	if _, err := bson.ObjectIDFromHex(id); err != nil {
		return 0, blogerrs.ErrInvalidPostID
	}
	return s.repo.IncrementLikes(ctx, id)
}

// ListTags 列出所有标签及文章数 —— 与 Python list_tags 对齐。
func (s *BlogService) ListTags(ctx context.Context) ([]dto.TagResponse, error) {
	tagCounts, err := s.repo.AggregateTagCounts(ctx)
	if err != nil {
		return nil, err
	}
	tags := make([]dto.TagResponse, 0, len(tagCounts))
	for _, tc := range tagCounts {
		tags = append(tags, dto.TagResponse{Name: tc.Name, Count: tc.Count})
	}
	return tags, nil
}

// ListPostsByTag 按标签分页列出博客 —— 与 Python get_posts_by_tag 对齐。
func (s *BlogService) ListPostsByTag(ctx context.Context, tag string, page, perPage int) (*dto.PostsByTagResponse, error) {
	tag = strings.TrimSpace(tag)
	if tag == "" {
		return nil, blogerrs.ErrInvalidPostID
	}
	if page < 1 {
		page = 1
	}
	if perPage < 1 {
		perPage = 10
	}

	posts, total, err := s.repo.ListPostsByTag(ctx, tag, page, perPage)
	if err != nil {
		return nil, err
	}

	return &dto.PostsByTagResponse{
		Posts:      dto.ToPostList(posts),
		Tag:        tag,
		Pagination: pagination(page, perPage, int(total)),
	}, nil
}
